back_end/src/models/friend.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+from typing import List
+
+from ..setup import db
+
+from .user import User
+
+
+class Friend(db.Model):
+    """
+    @author: YixuanZ
+    """
+    __tablename__ = "Friends"
+
+    id = db.Column(db.Integer, primary_key=True)
+    sender_id = db.Column(db.Integer, db.ForeignKey('Users.id'),
+                          nullable=False)
+    receiver_id = db.Column(db.Integer, db.ForeignKey('Users.id'),
+                            nullable=False)
+    accepted = db.Column(db.Boolean, nullable=False, default=False)
+
+    def __init__(self, **kwargs):
+        super(Friend, self).__init__(**kwargs)
+
+    def to_json(self):
+        ret = {}
+        ret['id'] = self.id
+        ret['sender_id'] = self.sender_id
+        ret['receiver_id'] = self.receiver_id
+        ret['accepted'] = self.accepted
+        return ret
+
+    def save(self):
+        db.session.commit()
+
+    @staticmethod
+    def add_friend(sender_id: int, receiver_id: int) -> (bool, Friend):
+        f = Friend.query.filter_by(sender_id=sender_id,
+                                   receiver_id=receiver_id).first()
+        if f:
+            return False, None
+        r = Friend.query.filter_by(sender_id=receiver_id,
+                                   receiver_id=sender_id).first()
+        if r:
+            return False, None
+        f = Friend(sender_id=sender_id, receiver_id=receiver_id,
+                   accepted=False)
+        db.session.add(f)
+        f.save()
+        return True, f
+
+    @staticmethod
+    def accept_friend(sender_id: int, receiver_id: int) -> (bool, str, Friend):
+        f = Friend.query.filter_by(sender_id=sender_id,
+                                   receiver_id=receiver_id).first()
+        if not f:
+            return False, "request not found", None
+        elif f.accepted:
+            return False, "request already accepted", f
+        else:
+            f.accepted = True
+            f.save()
+            return True, "success", f
+
+    @staticmethod
+    def is_friend(user1_id: int, user2_id: int) -> bool:
+        f = Friend.\
+            get_friend_by_sender_and_receiver(sender_id=user1_id,
+                                              receiver_id=user2_id).first()
+        r = Friend.\
+            get_friend_by_sender_and_receiver(sender_id=user2_id,
+                                              receiver_id=user1_id).first()
+        if f:
```

Comment:
potential bug here. In order to work we must assume that it is never the case where neither of f and r is none, i.e. it cannot happen that we have both relation A => B and B => A at the same time. 

It seems the code will work but think about what if both A and B send a request to each other at the same time. I have no idea how the synchronization is done in our database so It is definitely nice if there is some kind of lock that prevents this from happening, but if in the future something weird happens, this might need to be fixed. For now, I think it is ok.